enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,328 @@
+  - "/enhancements/update/cluster-profiles.md"
+---
+
+# Cluster Capabilities API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As we add more deployment topologies for OpenShift, we need a way to
+communicate expectations to the operators managing the cluster. This
+enhancement describes an API for expressing the expected
+"capabilities" of a cluster, including aspects such as how hard
+operators should work to provide high-availability and whether
+operators should configure their operands in a way to minimize
+resource consumption for especially constrained environments.
+
+## Motivation
+
+We want to avoid having cluster profiles use different manifests to
+configure operators managed by the cluster-version-operator, and
+instead rely on in-cluster APIs. This approach will make the same
+information available to operators managed by either the
+cluster-version-operator or operator-lifecycle-manager.
+
+We do not want to expose the cluster profile name directly to most of
+the operators, because we do not want to have to update operators each
+time we add a new profile.
+
+We want the settings to be global to the cluster, rather than using
+operator-specific APIs, because that simplifies the integration with
+the installer and allows them to be applied consistently in all
+operators that need them.
+
+### Goals
+
+- Expose capability expectations to operators running in a cluster so
+  they can configure their operands appropriately.
+
+### Non-Goals
+
+- Expose cluster profile names to operators.
+- Provide a mutable API for changing the cluster capabilities (at
+  least for now).
+- Predict other capability settings that we might need to expose.
+
+## Proposal
+
+A new [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations),
+`capabilities.config.openshift.io`, is created to hold details about
+the configured features of a cluster. It will include two fields to
+start, and may be expanded over time.
+
+The `highAvailabilityMode` field will use an enum to express the
+high-availability expectations. Possible values are `full` and
+`none`. The default is `full`, which represents the behavior operators
+have today in a "normal" cluster. The `none` mode will be used in
+single-node deployments (developer and production), and the operator
+should not configure the operand for highly-available
+operation. Systems such as the assisted installer, which may
+experience periods early in the lifetime of the cluster when full HA
+is not possible, can start with `none` and transition to `full` when
+the cluster is more fully online.
```

Comment:
This conflicts with the "Provide a mutable API for changing the cluster capabilities..." non-goal.  I think it's fine to drop that as a non-goal, and to have HA-ness in a `spec` property.  Then... something... maybe [the config operator][1] would validate incoming `spec` properties and pass valid choices through to `status` properties.  Other components would consume the `status` components (so the validation buffers them from a mass explosion if a user typos and edit to a `spec` property).  Once you have that sort of dynamic control, you probably want `status.conditions[]` as well, so the config operator has space to file complaints with details about any `spec` values that it is mad about.

[1]: https://github.com/openshift/cluster-config-operator

Reply:
Hmm yeah, supporting dynamic transitions is a whole other thing.  Is there a problem today with the assisted installer having operators in degraded mode until everything comes online?

An immutable property seems better to me.

Reply:
Currently, assisted-installer is using operator-specific APIs to tell the CEO to start the etcd cluster without waiting for 3 master nodes (the bootstrap will pivot to be the 3rd master node once the control plane is running on the first 2 masters).
The assisted-installer has a similar issue with kubernetes-OVN (which won't configure OVS without 3 master nodes (it actually looks at the number control plane replica in cluster-config configMap).
These components functionality is mandatory for the control plane to start on the 2 master nodes and for the installation to complete.

Reply:
I dropped the discussion of special handling for the assisted installer workflow and tried to emphasize more the immutable nature of the first version of the API.